File: iOS Examples/Objective-C/PlayAudioFile/PlayAudioFile/Instruments/AudioFilePlayer.m
//
//  AudioFilePlayer.m
//  AudioKit Example
//
//  Created by Aurelius Prochazka on 6/16/12.
//  Copyright (c) 2012 Aurelius Prochazka. All rights reserved.
//

#import "AudioFilePlayer.h"

@implementation AudioFilePlayer

- (instancetype)init {
    self = [super init];
    if (self) {
        
        // NOTE BASED CONTROL ==================================================
        AudioFilePlayerNote *note = [[AudioFilePlayerNote alloc] init];
        
        // INSTRUMENT DEFINITION ===============================================
        NSString *file = [[NSBundle mainBundle] pathForResource:@"blaster" ofType:@"aiff"];
        AKSoundFileTable *fileTable = [[AKSoundFileTable alloc] initWithFilename:file];
        
        AKMonoSoundFileLooper *looper;
        looper = [AKMonoSoundFileLooper looperWithSoundFile:fileTable];
        looper.frequencyRatio = note.speed;
        looper.loopMode = [AKMonoSoundFileLooper loopPlaysOnce];
        
        AKReverb *reverb;
        reverb = [AKReverb reverbWithInput:looper];
        reverb.feedback.value = 0.85;
        
        [self setStereoAudioOutput:reverb];
    }
    return self;
}

@end

// -----------------------------------------------------------------------------
#  pragma mark - Instrument Note
// -----------------------------------------------------------------------------

@implementation AudioFilePlayerNote

- (instancetype)init;
{
    self = [super init];
    if(self) {
        _speed = [self createPropertyWithValue:1.0 minimum:0.2 maximum:2.0];
    }
    return self;
}
- (instancetype)initWithSpeed:(float)speed;
{
    self = [self init];
    if(self) {
        self.speed.value = speed;
    }
    return self;
}

@end

File: iOS Examples/Swift/PlayAudioFile/PlayAudioFile/AudioFilePlayer.swift
//
//  AudioFilePlayer.swift
//  PlayAudioFile
//
//  Created by Matthew Fecher on 5/23/15.
//  Copyright (c) 2015 audiokit.io. All rights reserved.
//

class AudioFilePlayer: AKInstrument {

    override init() {
        super.init()
        
        // NOTE BASED CONTROL ==================================================
        let note = AudioFilePlayerNote()
        
        // INSTRUMENT DEFINITION ==================================================
        let file = NSBundle.mainBundle().pathForResource("blaster", ofType: "aiff")
        let fileTable = AKSoundFileTable(filename: file)
        
        let looper = AKMonoSoundFileLooper(soundFile: fileTable)
        looper.frequencyRatio = note.speed
        looper.loopMode = AKMonoSoundFileLooper.loopPlaysOnce()
        
        // EFFECT
        let reverb = AKReverb(input: looper)
        reverb.feedback.value = 0.85
        
        // AUDIO OUTPUT ========================================================
        setStereoAudioOutput(reverb)
    }
}

// -----------------------------------------------------------------------------
// MARK: - Instrument Note
// -----------------------------------------------------------------------------

class AudioFilePlayerNote: AKNote {
    
    // Note Properties
    var speed = AKNoteProperty(value: 1.0, minimum: 0.2, maximum: 2.0)
    
    override init() {
        super.init()
        addProperty(speed)
    }
    
    convenience init(speed: Float) {
        self.init()
        self.speed.value = speed
    }
}
